### PR TITLE
Several breaking changes to the helm chart

### LINKED ISSUE #635: Don't install snapshot controller
/kind bug

**What happened?** Snapshot controller is installed by helm.

**What you expected to happen?** Snapshot controller should not be installed, the driver should reuse an existing installation of snapshot controller, since the controller might be reused by other drivers. Or helm could somehow check if the controller already exists. Trying to install it again can cause conflict issues.

**How to reproduce it (as minimally and precisely as possible)?** do helm install

**Anything else we need to know?**: see also https://github.com/kubernetes-sigs/aws-ebs-csi-driver/pull/601

**Environment**
- Kubernetes version (use `kubectl version`):
- Driver version:

### LINKED ISSUE #847: extraVolumeTags chart templating type format issue
/kind bug

**What happened?**
Extra-volume-tags with a value of a type other than string are not correctly formatted.

**How to reproduce it (as minimally and precisely as possible)?**
Here is an example:
```
➜  ~ helm install aws-ebs-csi-driver aws-ebs-csi-driver/aws-ebs-csi-driver --set extraVolumeTags.astring=foo --set extraVolumeTags.abool=true --set extraVolumeTags.anumber=22 --dry-run | grep extra-volume-tags
            - --extra-volume-tags=abool=%!s(bool=true),anumber=%!s(int64=22),astring=foo
```

**Proposal**
Tags formating [here](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/57e57bb10bd5e7f52596b673d61ce45d751e3025/charts/aws-ebs-csi-driver/templates/_helpers.tpl#L64) should be changed this way `{{- $noop := printf "%s=%v" $key $value | append $result.pairs | set $result "pairs" -}}`

**Environment**
- Kubernetes version (use `kubectl version`): not related
- Driver chart version: 0.10.2

### LINKED ISSUE #964: Breaking changes for v2.0.0 of the helm chart
/kind feature

**Describe the solution you'd like in detail**
There are deprecated values in the helm chart.  They should be removed in the next major release of the chart.
The helm chart still supports helm v2 which was end of lifed in November 2020.  The next major release should require helm v3 or above.

**Additional context**
PR #856 deprecated the fields in favor of a more clear organization of values.

### PR DESCRIPTION
**Pull request by krmichel**
**Is this a bug fix or adding new feature?**
Both and includes a breaking change

Fixes #635 
Fixes #847 
Fixes #964 

**What is this PR about? / Why do we need it?**
Removes deprecated helm values
Removes support for helm v2
Removes installation of the [Snapshot controller](https://github.com/kubernetes-csi/external-snapshotter)

### CODE REVIEW

**Comment by krmichel** [comment]
@wongma7 Here is the PR I talked about with the breaking changes.

**Comment by krmichel** [comment]
I will need to add some information to the readme about these changes and upgrading from chart version 1.X to 2.X.

**Comment by krmichel** [comment]
Looks like the tests are failing since the CSI Snapshotter is no longer being installed.  I will try taking a look later.

**Comment by wongma7** [comment]
https://github.com/kubernetes/kops/blob/ceb5031b59f2408b468ac07a6e38a1fab83e8842/docs/addons.md#snapshot-controller

ah, it only gets auto-installed in 1.21 not 1.20... Our tests are creating 1.20 kops clusters. so, still have to set 
spec:
  snapshotController:
    enabled: true
in https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/master/hack/kops-patch.yaml#L1

alternatively, kops 1.21 just came out a few days ago, so our tests could create 1.21 kops clusters, but that might be out of scope of this PR https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/master/hack/e2e/run.sh#L51-L55

File: charts/aws-ebs-csi-driver/templates/controller.yaml [hunk 1]
```diff
@@ -185,7 +182,7 @@ spec:
           volumeMounts:
             - name: socket-dir
               mountPath: /var/lib/csi/sockets/pluginproxy/
-          {{- with default .Values.resources (default .Values.controller.resources .Values.controller.containerResources.attacher) }}
+          {{- with default .Values.controller.resources .Values.controller.containerResources.attacher }}
           resources:
             {{- toYaml . | nindent 12 }}
           {{- end }}
```
Comment (wongma7) [inline charts/aws-ebs-csi-driver/templates/controller.yaml:188]: IMO we should remove the wrapper around csi-snapshotter and  remove `enableVolumeSnapshot`

As long as we document that having the CRDs and snapshot-controller is a prerequisite, and IMO it makes sense to include the same note in README as in NOTES.txt, users should not be surprised to see csi-snapshotter is failing.

(Sorry to sound like a broken record on this issue, really want to be rid of it!)

**Comment by wongma7** [comment]
FYI I intend to add a helm changelog (copying how efs did it) here : https://github.com/kubernetes-sigs/aws-ebs-csi-driver/pull/958/files#diff-3e7c1a6135a81558fa6110f37850a0aea8e210a2226bd7c35c3f0819e5a74fe1 . I am hoping I can get it merged soon, in which case you may need to rebase

**Comment by wongma7** [comment]
As for the EKS test failure, that's due to https://github.com/kubernetes-sigs/aws-ebs-csi-driver/pull/967, we can ignore it.

**Comment by krmichel** [comment]
> https://github.com/kubernetes/kops/blob/ceb5031b59f2408b468ac07a6e38a1fab83e8842/docs/addons.md#snapshot-controller
> 
> ah, it only gets auto-installed in 1.21 not 1.20... Our tests are creating 1.20 kops clusters. so, still have to set
> spec:
> snapshotController:
> enabled: true
> in https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/master/hack/kops-patch.yaml#L1
> 
> alternatively, kops 1.21 just came out a few days ago, so our tests could create 1.21 kops clusters, but that might be out of scope of this PR https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/master/hack/e2e/run.sh#L51-L55

I might be misreading but I think we will need kops 1.21 but can continue with kube 1.20.  I am basing that on [this representation](https://kops.sigs.k8s.io/addons/#snapshot-controller) seeming more clear to me.  I will try it when I get a chance

**Comment by wongma7** [comment]
You are right, I misunderstood relationship between kops version and k8s version.

I just tried kops 1.20.x to create a 1.20.x k8s version and it ignored snapshotController.

Let me try kops 1.21.x to create a 1.20.x: https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/master/hack/e2e/run.sh#L55

File: charts/aws-ebs-csi-driver/values.yaml [hunk 2]
```diff
@@ -84,13 +61,13 @@ controller:
...
+  priorityClassName: system-cluster-critical
   # AWS region to use. If not specified then the region will be looked up via the AWS EC2 metadata
   # service.
   # ---
   # region: us-east-1
   region:
-  replicaCount:
+  replicaCount: 2
   resources: {}
   containerResources:
```
Comment (wongma7) [inline charts/aws-ebs-csi-driver/values.yaml:72]: WDYT about renaming the sidecars.*image keys to be sidecar.*.image, and then adding sidecar.*.resources, to be consistent with EFS? https://github.com/kubernetes-sigs/aws-efs-csi-driver/blob/master/charts/aws-efs-csi-driver/values.yaml#L21

File: charts/aws-ebs-csi-driver/values.yaml [hunk 3]
```diff
@@ -146,17 +119,14 @@ node:
     ebsPlugin: {}
     nodeDriverRegistrar: {}
     liveness: {}
+  # The "maximum number of attachable volumes" per node
   volumeAttachLimit:
 
 serviceAccount:
```
Comment (wongma7) [inline charts/aws-ebs-csi-driver/values.yaml:125]: can we move it to under controller.serviceAccount and node.serivceAccount (again, to be consistent with EFS https://github.com/kubernetes-sigs/aws-efs-csi-driver/blob/master/charts/aws-efs-csi-driver/values.yaml#L69).

File: hack/values_eksctl.yaml [hunk 4]
```diff
@@ -1,4 +1,3 @@
-enableVolumeSnapshot: true
 controller:
   logLevel: 5
 node:
```
Comment (wongma7) [inline hack/values_eksctl.yaml:3]: the serviceAccount.controller.create here also needs shuffling to controller.serviceAccount.create  : x

File: docs/README.md [hunk 5]
```diff
@@ -176,6 +177,73 @@ helm upgrade --install aws-ebs-csi-driver \
     aws-ebs-csi-driver/aws-ebs-csi-driver
 ```
 
+##### Upgrading from version 1.X to 2.X of the helm chart
```
Comment (wongma7) [inline docs/README.md:180]: could you also copy/paste or link this in the new CHANGELOG.md? https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/master/charts/aws-ebs-csi-driver/CHANGELOG.md

(I will edit our github release workflow to link to the CHANGELOG )

**Comment by wongma7** [comment]
/lgtm

thanks very much.

/hold

last chance for folk to mention any 2.0 changes they would like to see...

File: docs/README.md (hunk 5 above)
Comment (krmichel) [reply docs/README.md:180]: Done

File: charts/aws-ebs-csi-driver/values.yaml [hunk 6]
```diff
@@ -84,21 +78,18 @@ controller:
   nodeSelector: {}
   podAnnotations: {}
   podLabels: {}
-  priorityClassName:
```
Comment (wongma7) [inline charts/aws-ebs-csi-driver/values.yaml:87]: the `extraVars` value above can be removed

File: charts/aws-ebs-csi-driver/templates/controller.yaml [hunk 7]
```diff
@@ -97,7 +94,7 @@ spec:
                   name: aws-secret
                   key: access_key
                   optional: true
-            {{- with default .Values.region .Values.controller.region }}
+            {{- with .Values.controller.region }}
             - name: AWS_REGION
               value: {{ . }}
             {{- end }}
```
Comment (wongma7) [inline charts/aws-ebs-csi-driver/templates/controller.yaml:100]: extraVars here can be removed

**Comment by wongma7** [comment]
/lgtm
/hold cancel

thank you!!!

**Pull request merged**